Make knn search a query

mayya-sharipova opened the pull request:
This adds a new knn query:
- knn query is executed during the Query phase similar to all other queries.
- No k parameter, k defaults to  size
- num_candidates - number of closest neighbours that knn query returns on each shard
- For aggregations: "size" results are collected with total = size * shards. Aggregations will see size * shards results.
- All filters from DSL are applied as post-filters, except: 1) alias filter is applied as  pre-filter or 2) a filter provided as a parameter inside knn query.

Closes #97940

mayya-sharipova committed:
Make knn search a query

This introduced a new knn query:
- knn query is executed during the Query phase similar to all other queries.
- No k parameter, k defaults to  size
- num_candidates is a size of queue for candidates to consider while
  search a graph on each shard
- For aggregations: "size" results are collected with total = size * shards.
   Aggregations will see size * shards results.
- All filters from DSL are applied as post-filters, except: 1) alias filter
 is applied as  pre-filter or 2) a filter provided as a parameter
 inside knn query.

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

mayya-sharipova committed:
Update docs/changelog/98916.yaml

elasticsearchmachine commented:
Hi @mayya-sharipova, I've created a changelog YAML for you.

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/main' into knn-as-query

mayya-sharipova committed:
Correct transport version, remove byteQueryVector

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/main' into knn-as-query

mayya-sharipova committed:
Other adjustments

benwtrent commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1252,6 +1252,11 @@ private void parseSource(DefaultSearchContext context, SearchSourceBuilder sourc
         QueryBuilder query = source.query();
         if (query != null) {
             InnerHitContextBuilder.extractInnerHits(query, innerHitBuilders);
+            // apply alias filter to as preFilter to queries that need it (e.g. knn)
+            QueryBuilder aliasFilter = context.request().getAliasFilter().getQueryBuilder();
+            if (aliasFilter != null) {
+                query.addFilterQuery(aliasFilter);
+            }
```
What do you think about adding the `aliasFilter` to the `SearchExecutionContext` ?

Earlier in this method:
```
SearchExecutionContext searchExecutionContext = context.getSearchExecutionContext();
```

During rewrite the knnQueryBuilder could check `searchExecutionContext#getAliasFilter()` and apply the filters internally.

This seems simpler than adding a new method to all query builder objects (which doesn't seem strictly required to me).

mayya-sharipova replied:
That's a great suggestion, the code change is much smaller now. Addressed in 9b1e7c86bea2e72e23b24245cabafe1ff3c823d9

benwtrent commented on server/src/main/java/org/elasticsearch/search/vectors/KnnVectorQueryBuilder.java:
```diff
@@ -23,56 +24,131 @@
+    public static final ParseField NUM_CANDS_FIELD = new ParseField("num_candidates");
+    public static final ParseField QUERY_VECTOR_FIELD = new ParseField("query_vector");
+    public static final ParseField QUERY_VECTOR_BUILDER_FIELD = new ParseField("query_vector_builder");
+    public static final ParseField VECTOR_SIMILARITY_FIELD = new ParseField("similarity");
```
I am not 100% sure we should have `similarity` here in the `knn` query object.

The reason it was added to the top level `knn` object was because we couldn't have a `vectorSimilarity` query. But now we have opened the flood gates for allowing more than one kind of vector query.

Do we think we should have `similarity` here or not?

mayya-sharipova replied:
@benwtrent Thanks for the comment. 
In my opinion, it seems excessive to introduce a new `vectorSimilarity` query just to limit scores of a knn query.
I think we either should:
1. Remove `similarity` from a knn query, and ask users to use `min_score` search request parameter. But, the user should provide minimum score (not similarity)
2. Keep `similarity` parameter in the query. 

WDYT of these options?

benwtrent replied:
`min_score` in the search request applies to the entire search clause. It seems like it misses a key usage where `knn` is used within a `filter` clause and executed with aggregations. I don't think we can use the search request `min_score` as a substitute.

I also see that `function_score` has a `min_score` parameter :(. But that thing is so complicated.

So, my counter options would be:
 - keep it in the query
 - we remove it and add a separate one called `vectorSimilarity`.

I am just trying to think of what we would have done if `knn` was always a query and we added similarity thresholds later. Would we have made it a separate query or added a parameter?

jdconrad replied:
I agree that `min_score` doesn't seem like a good candidate since it's global within a search. If we continue down the path of additional queries to simplify the API then it makes sense to me to have a wrapper for either a min_score query in general or a specific threshold query to wrap a knn query.

mayya-sharipova replied:
I am ok with having an extra `vector_similarity` query that wraps a knn query or `semantic_search` query (a new query that we will introduce for query_vector_builder.

```js
  "vector_similarity": {
    "knn": {
	    "field": "dense-vector-field",
	    "num_candidates": 100,
	    "query_vector" : [...],
	    "text": "The opposite of blue"
  	},
  	"similarity" : 10
  }
```

```js
  "vector_similarity": {
    "semantic_search": {
	    "field": "dense-vector-field",
	    "num_candidates": 100,
	    "model_id": "my-text-embedding-model",
	    "text": "The opposite of blue"
  	},
  	"similarity" : 10
  }
```

mayya-sharipova replied:
@giladgal What would be your opinion?

giladgal replied:
I don't think there is an obvious correct choice between these two options. Having a parent query seems more organized and then that same query could be used with different types of vector queries, as long as you know which queries it can be used with. 

Personally I prefer a parameter than a parent query. I think having all vectors being candidates for any query vector is not how we humans think of similarity and relevance. When the engine returns irrelevant results just because they are the most similar results, it is probably not what the end user wanted. In other words I see the similarity threshold as a natural part of a vector search and I would like for it to be documented so that anyone that reads about vector search immediately becomes aware of that important option. I also like that a parameter is simpler to add and I find it easier to read.

benwtrent commented on server/src/main/java/org/elasticsearch/search/vectors/KnnVectorQueryBuilder.java:
```diff
@@ -23,56 +24,131 @@
+    public static final ParseField FIELD_FIELD = new ParseField("field");
+    public static final ParseField NUM_CANDS_FIELD = new ParseField("num_candidates");
+    public static final ParseField QUERY_VECTOR_FIELD = new ParseField("query_vector");
+    public static final ParseField QUERY_VECTOR_BUILDER_FIELD = new ParseField("query_vector_builder");
```
I think we shouldn't have `query_vector_builder` in the knn query. 

It seems like we should instead of a `semanticVectorQuery` that rewrites itself to a `knn` query after fetching the correct information.

We were hamstrung by the current top level `knn` design.

mayya-sharipova replied:
Thanks for the feedback. Are you suggesting to introduce a new query like this?

```
"semantic_search": {
    "field": "dense-vector-field",
    "num_candidates": 100,
    "query_vector_builder": {
      "text_embedding": { 
        "model_id": "my-text-embedding-model", 
        "model_text": "The opposite of blue" 
      }
    }
  }
```
What are your reasons why we can't have `query_vector_builder` inside a knn query?  There would be a lot of duplication for  `semantic_query`.

benwtrent replied:
> What are your reasons why can't we have query_vector_builder inside a knn query? There would be a lot of duplication for semantic_query.

It complicates the query. We have rewrite phases, etc. to make this sort of thing simpler. 

I think that a "semantic_search" query can simplify things and is the correct way to handle this. 

It could even dynamically handle switching between `dense` and `sparse` models.

IMO A semantic search query would be something like
```
"semantic_search": {
    "field": "dense-vector-field",
    "num_candidates": 100,
    "model_id": "my-text-embedding-model",
    "text": "The opposite of blue"
  }
```

mayya-sharipova replied:
Thanks for further clarifying, Ben! I will remove `query_vector_builder` from knn query.

benwtrent reviewed (commented):
I have a concern about inheriting designs that were required because of how the top-level `knn` was. I think we can do better here when adding a new query.

mayya-sharipova committed:
Add aliasFilter to the SearchExecutionContext instead of QueryBuilder

mayya-sharipova committed:
Remove query_vector_builder

benwtrent commented:
@mayya-sharipova could you add a test that ensures the `knn` query supports `_name`? It would be good to know which `knn` vectors contributed to a score or not. The top level `knn` currently doesn't support this.

jdconrad commented on server/src/main/java/org/elasticsearch/index/query/SearchExecutionContext.java:
```diff
@@ -228,6 +230,15 @@ private void reset() {
         this.nestedScope = new NestedScope();
     }
 
+    // Set alias filter, so it can be applied for queries that need it (e.g. knn query)
+    public void setAliasFilter(QueryBuilder aliasFilter) {
+        this.aliasFilter = aliasFilter;
+    }
+
+    public QueryBuilder getAliasFilter() {
+        return aliasFilter;
+    }
+
```
I wonder if `SearchContext` should be made available as part of `rewrite` instead of `SearchExecutionContext` since alias filter is available as part of that? I do concede this may be too large a change for now.

mayya-sharipova replied:
Thanks for the feedback, but looks like indeed this would be too large a change

jdconrad reviewed (commented):
This is a great change! I added a couple of thoughts inline.

benwtrent commented on server/src/main/java/org/elasticsearch/search/vectors/KnnVectorQueryBuilder.java:
```diff
@@ -199,11 +212,14 @@ public String getWriteableName() {
-    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws IOException {
+    protected QueryBuilder doRewrite(QueryRewriteContext ctx) throws IOException {
         boolean changed = false;
+        if (ctx instanceof SearchExecutionContext searchExecutionContext && searchExecutionContext.getAliasFilter() != null) {
+            filterQueries.add(searchExecutionContext.getAliasFilter());
+        }
```
How does this rewrite ever exit and do we continually append the alias filter on subsequent rewrites?

What do you think about adding this code to `doToQuery` instead of rewrite? It seems to me that the aliasFilter will be rewritten already and then when we call `doToQuery` we take the filter and append it to the list (not updating the query object, but instead adjusting the boolean query directly)

Like:
```
        BooleanQuery.Builder builder = new BooleanQuery.Builder();
        for (QueryBuilder query : this.filterQueries) {
            builder.add(query.toQuery(context), BooleanClause.Occur.FILTER);
        }
        if (searchExecutionContext.getAliasFilter() != null) {
            builder.add(searchExecutionContext.getAliasFilter().toQuery(context), BooleanClause.Occur.FILTER);
        }
        BooleanQuery booleanQuery = builder.build();
        Query filterQuery = booleanQuery.clauses().isEmpty() ? null : booleanQuery;

```

mayya-sharipova replied:
Ben, indeed I have not thought through about rewrite. Great suggestion about putting into `doToQuery`, I confirm that alias filter  is already rewritten by this stage.  Addressed in 8a4cb9b7922372317a7d01594ab85856e2ce317e.

mayya-sharipova committed:
Add query _name to tests

mayya-sharipova committed:
Add filter alias during doToQuery

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/main' into knn-as-query

mayya-sharipova committed:
Simplify over-wire protocol

mayya-sharipova commented:
> @mayya-sharipova could you add a test that ensures the knn query supports _name? It would be good to know which knn vectors contributed to a score or not. The top level knn currently doesn't support this.

@benwtrent  Great feedback, indeed interesting and necessary test cases, addressed in 9046812f95bf55778897099c783f9e846328bfef

mshameti commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/90_knn_query_with_filter.yml:
```diff
@@ -0,0 +1,253 @@
+                    field: my_vector
+                    query_vector: [ 1, 1, 1, 1 ]
+                    num_candidates: 5
+                    filter:
```
Speaking purely from the viewpoint of a consumer of this API without deep knowledge of the underlying implementation of kNN search in ES:

In the spirit of simplifying the API, is there an opportunity to a) remove the 'filter' field from the knn query and b) have the kNN query be the post-filter instead?  

Assuming the current spec changes, if I were to provide a 'term' and a 'knn' clause then I am actually more likely to see fewer results than if I were to provide a 'term' clause and a 'knn' clause with a 'filter', due to kNN trying to satisfy the 'size' requirement as quickly as possible.

More clauses yielding more results feels counter-intuituive in a context other than 'should'

mayya-sharipova replied:
@mshameti Thanks for your feedback.

Are you suggesting to  always exercise filters for knn query as post filters?
Pre-filtering is very important aspect of knn query that ensures you will get some results, while post-filter can eliminate all results.

> 'term' and a 'knn' clause then I am actually more likely to see fewer results than if I were to provide a 'term' clause and a 'knn' clause with a 'filter', due to kNN trying to satisfy the 'size' requirement as quickly as possible.

Not sure how is it more likely?  It is actually less likely. `knn` with a `filter` returns the same or less number of results than knn query, but never more.

mshameti replied:
Thanks for looking! 

I've created a Gist to demonstrate this a bit better hopefully. I've listed out my assumptions in there as well.

 https://gist.github.com/mshameti/774639631363457dec310deee4f15766

benwtrent replied:
> More clauses yielding more results feels counter-intuitive in a context other than 'should'

It doesn't return more results. `knn` will apply the filter and when combined with other queries, those may be "post filtered" out. The doc count hit may still be `k` as we found there are at least `k` documents that match that filter and scored them.

But the scored hits returned must satisfy all query provided filters. I am not sure how it returns MORE results. Reading through the gist, I am still not 100% sure I understand the concern.

Could you clarify what you mean by "More clauses yielding more results feels counter-intuitive in a context other than 'should'"? 

The scenario with a `must`

```
GET /knn-test/_search
{
  "query": {
    "bool": {
      "must": [
        {
          "match": {
            "author": "Martin"
          }
        },
        {
          "knn": {
            "query_vector": [1, 1, 1, 1],
            "field": "paragraph_embedding",
            "num_candidates": 1,
            "filter": {
              "match": {
                "author": "Martin"
              }
            }
          }
        }
      ]
    }
  },
  "size": 1
}
```

This would result in only documents that satisfy 
```
{
          "match": {
            "author": "Martin"
          }
        }
```

And `knn` will only score documents where
```
"filter": {
              "match": {
                "author": "Martin"
              }
            }
```

Is true.

So, it isn't finding MORE documents. It is just scoring different documents, which still matches the typical way a query works, its just that other queries have a "natural pre-filter" which is the matching terms :).

benwtrent replied:
> a. treat the top level match as a pre-filter

I understand now. 

We considered that, but it gets very difficult to get correct and would have non-obvious edge cases. 

What do you do in a multiple nested `bool` query?

What if a higher level `bool` has a `knn` query? 

Do we ignore `should` clauses?

We may in the future consider making the filter optional and dynamically applying the correct or expected pre-filter.

mshameti replied:
Great questions! For future use, I created a Gist of the situtations you mentioned and a couple others to help me reason through this. https://gist.github.com/mshameti/2b87802869d83df6b06abade142a922e

I most likely have not covered all the bases but if you do decide to revisit this it's here for future use. 

Thanks again for engaging in the discussion, and as the proposed API change is already an improvement to the existing API, please consider this conversation resolved from my end!

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/main' into knn-as-query

benwtrent commented on server/src/main/java/org/elasticsearch/search/vectors/KnnVectorQueryBuilder.java:
```diff
@@ -28,74 +28,102 @@
+    public static final ParseField FIELD_FIELD = new ParseField("field");
+    public static final ParseField NUM_CANDS_FIELD = new ParseField("num_candidates");
+    public static final ParseField QUERY_VECTOR_FIELD = new ParseField("query_vector");
+    public static final ParseField VECTOR_SIMILARITY_FIELD = new ParseField("similarity");
```
I am leaning towards just a `vector_similarity` query, not a parent query, etc. and  not using this parameter when used as a `query`.

mayya-sharipova replied:
If we want to remove `similarity` field, we need to introduce another QueryBuilder separate from the current  `KnnVectorQueryBuilder`that is used by knn search and uses `similarity` field.

benwtrent replied:
@mayya-sharipova I am ok having it their for now. It is simplest and keeps uniformity with top-level knn. 

If we ever introduce a `vector_similarity_query`, we would deprecate this parameter and point folks to use it instead.

mayya-sharipova committed:
Add nested support for knn query

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/main' into knn-as-query

mayya-sharipova commented:
@benwtrent  @jdconrad This is ready for another round of review with nested support added.

benwtrent commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/110_knn_query_with_filter.yml:
```diff
@@ -0,0 +1,253 @@
+  - match: { hits.hits.2._id: "3" }
+  - match: { hits.hits.2.fields.my_name.0: v1 }
+---
+"PRE_FILTER: knn query with alias filter as pre-filter":
```
IT JUST WORKS! I love this.

benwtrent commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/130_knn_query_nested_search.yml:
```diff
@@ -0,0 +1,212 @@
+                        query_vector: [ -0.5, 90.0, -10, 14.8, -156.0 ]
+                        num_candidates: 6
+              inner_hits: { size: 1, "fields": [ "nested.paragraph_id" ], _source: false }
+  # no hits because, regardless of num_candidates knn returns top 3 child vectors from distinct parents
+  # and they don't pass the post-filter
+  # TODO: fix it on Lucene leve so nested knn respects num_candidates
+  - match: {hits.total.value: 0}
```
This is indeed tricky. It does satisfy `num_candidates`, but how we are pre-filtering things can act weird with other nested queries. You are correct, it is worth revisiting in the future.

Along with allowing nested pre-filtering (filtering on child field values instead of just parent values).

benwtrent commented on rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/search.vectors/40_knn_search.yml:
```diff
@@ -294,23 +294,6 @@ setup:
   - match: { error.root_cause.0.reason: "failed to create query: field [nonexistent] does not exist in the mapping" }
 
 ---
-"Direct kNN queries are disallowed":
```
:godmode:

benwtrent reviewed (approved):
This makes me so happy!

🎉 🎉 🎉 🎉 🎉 

A couple of things we should make sure of:

 - Does this work with `function_score` `dis_max` etc.? I am think it should without issue, but we should make 100% sure.
 - How about `percolate`?
 - Does `pinned` work? (I would think so, since we just rewrite to the scoring docs on the shard...).
 - We need to add `query-dsl-knn-query.asciidoc` and put it under "specialized queries".

mayya-sharipova committed:
Add documentation and other queries

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/main' into knn-as-query

mayya-sharipova commented:
@benwtrent Thanks for your review so far.

> A couple of things we should make sure of:
Does this work with function_score dis_max etc.? I am think it should without issue, but we should make 100% sure.
How about percolate?
Does pinned work? (I would think so, since we just rewrite to the scoring docs on the shard...).
We need to add query-dsl-knn-query.asciidoc and put it under "specialized queries".

With the latest [commit](https://github.com/elastic/elasticsearch/pull/98916/commits/b3ca3112b0ca3c81a689b82628bd78a6c8bbb026), I've do the following:
- Added more tests with other queries such as `function_score`, `dis_max` and `pinned` queries.
- Modified `percolate` query NOT to accept `knn` as an internal query.  I think it does not make sense to use percolate a document against `knn` query, as `knn` query matches any single document.  What do you think?
- Added documentation for a new knn query and also put this query under "specialized queries".

benwtrent reviewed (approved):
Some minor things.

benwtrent commented on modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorFieldMapper.java:
```diff
@@ -438,6 +439,8 @@ static QueryBuilder parseQueryBuilder(DocumentParserContext context) {
                     throw new IllegalArgumentException("the [has_child] query is unsupported inside a percolator query");
                 } else if (queryName.equals("has_parent")) {
                     throw new IllegalArgumentException("the [has_parent] query is unsupported inside a percolator query");
+                } else if (queryName.equals(KnnVectorQueryBuilder.NAME)) {
+                    throw new IllegalArgumentException("the [knn] query is unsupported inside a percolator query");
```
>  What do you think? By default, knn and percolate don't work together and this requires more investigation. But more importantly, I think it does not make sense to percolate a document against knn query, as knn query matches any single document.

I think this is OK for now. 

I will have to think more about if we should support it or not. IMO, it seems like we should. We are just matching the nearest "k", so it seems to fit OK. But I can also see the argument against (as you laid out).

Additionally, we should have a "more_like_this" query that utilizes knn as well. 

Maybe open a Github issue to track discussion?

mayya-sharipova replied:
This as well, but more importantly, I think it does not make sense to percolate a document against knn query, as knn query matches any single document.

mayya-sharipova replied:
@benwtrent Let's discuss this offline (about percolator and more_like_this queries) and create github issues if we find them necessary.

I will consider this PR will go without those queries.

mayya-sharipova committed:
Updates to knn-query documentation

mayya-sharipova commented on docs/reference/query-dsl/knn-query.asciidoc:
```diff
@@ -0,0 +1,238 @@
+all documents are allowed to match.
+
+The filter is a pre-filter, meaning that it is applied **during** the approximate
+kNN search to ensure that `k` matching documents are returned.
```
Addressed in https://github.com/elastic/elasticsearch/commit/9d825cf183f557cde313d840fe0be2c094b25585

mayya-sharipova commented on docs/reference/query-dsl/knn-query.asciidoc:
```diff
@@ -0,0 +1,238 @@
+
+NOTE: `knn` query doesn't have a separate `k` parameter. `k` is defined by
+`size` parameter of a search request similar to other queries. `knn` query
+collects num_candidates` results from each shard, then merges them to get
```
Addressed in https://github.com/elastic/elasticsearch/commit/9d825cf183f557cde313d840fe0be2c094b25585

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/main' into knn-as-query

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/main' into knn-as-query

mayya-sharipova commented:
@elasticmachine run elasticsearch-ci/docs

mayya-sharipova commented:
@elasticmachine run elasticsearch-ci/docs

mayya-sharipova committed:
Adjust docs

mayya-sharipova committed:
Merge remote-tracking branch 'upstream/main' into knn-as-query

mayya-sharipova committed:
Merge branch 'main' into knn-as-query

mayya-sharipova committed:
Fix an error in docs

mayya-sharipova commented on docs/reference/query-dsl/knn-query.asciidoc:
```diff
@@ -0,0 +1,223 @@
+  }
+}
+----
+--
```
@abdonpijpelink Thank you.
